### PR TITLE
For 82

### PR DESCRIPTION
**Pull request by pniel-cohen**

### CODE REVIEW

File: view/frontend/templates/forter.phtml [hunk 1]
```diff
@@ -15,8 +33,37 @@ require(["jquery"], function ($) {
     } else {
       console.log('Forter: Invalid Web Id');
     }
+    });
+
+    <?php if ($preAuthIsActive && strlen($elementToObserve) > 1): ?>
+    $(document).on('change', '<?php echo $elementToObserve; ?>', function() {
+      value = $(this).val();
+      if(value.length === 16){
```
Comment (erezkoren) [inline view/frontend/templates/forter.phtml:41]: credit cards can be between 13-19 digits included

File: view/frontend/templates/forter.phtml [hunk 2]
```diff
@@ -15,8 +33,37 @@ require(["jquery"], function ($) {
...
+        var bin = '';
+        var last4cc = '';
+
+        var xhr = new XMLHttpRequest();
+        xhr.open("POST", '/forter/index/sessions', true);
```
Comment (erezkoren) [inline view/frontend/templates/forter.phtml:46]: this endpoint should not be related to forter --> it is better to change it to cardDetails endpoint or something similar.
we want the customer/developer will think the merchant is collecting the info of the bin+last4 and not the merchant

File: view/frontend/templates/forter.phtml [hunk 3]
```diff
@@ -15,8 +33,37 @@ require(["jquery"], function ($) {
     } else {
       console.log('Forter: Invalid Web Id');
     }
+    });
+
+    <?php if ($preAuthIsActive && strlen($elementToObserve) > 1): ?>
```
Comment (erezkoren) [inline view/frontend/templates/forter.phtml:38]: it is not necessarily related to pre-auth, we can use the bin+last4 also in postAuth

File: etc/adminhtml/system.xml [hunk 4]
```diff
@@ -193,7 +193,39 @@
                     <label>Enable Phone Order</label>
                     <source_model>Magento\Config\Model\Config\Source\Yesno</source_model>
                 </field>
-            </group>
+              </group>
+              <group id="advanced_settings_pre_auth" translate="label" type="text" sortOrder="40" showInDefault="1" showInWebsite="1" showInStore="1">
+                    <label>Advanced Settings Pre Auth</label>
```
Comment (erezkoren) [inline etc/adminhtml/system.xml:198]: it is not necessarily only in preauth, we want the capability and data also in the postauth

File: Block/Forter.php [hunk 5]
```diff
@@ -37,10 +36,48 @@ public function __construct(
...
     public function getSiteId()
     {
         return $this->forterConfig->getSiteId();
     }
+
+    /**
+     * Return boolean regarding active/disable pre-auth card observing
+     * @return boolean
+     */
+    public function getPreAuthIsActive()
```
Comment (orpolaczek) [inline Block/Forter.php:49]: not super critical, but usually this method would be called `isPreAuthActive`
If it's a must to have `get` before it, then `getIsPreAuthActive` :) 

Just makes the code more readable, and when I see `getIs........` in an if clause I know immediately it's a boolean

File: Controller/Index/Sessions.php [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        return parent::__construct($context);
+    }
+
+    /**
+     *
+     */
+    public function execute()
+    {
+        try {
+            $forterToken = $this->getRequest()->getHeader('Forter-Token');
```
Comment (orpolaczek) [inline Controller/Index/Sessions.php:42]: @zachGirit - IIRC we already added it. where did it go?

File: Model/RequestBuilder/BasicInfo.php [hunk 7]
```diff
@@ -31,10 +34,12 @@ public function __construct(
      */
     public function getConnectionInformation($remoteIp, $headers)
     {
+        $forterToken = $this->customerSession->getForterToken() ? $this->customerSession->getForterToken() : $this->cookieManager->getCookie("forterToken");
```
Comment (orpolaczek) [inline Model/RequestBuilder/BasicInfo.php:37]: same here. we already added this stuff here: https://github.com/forter/forter-magento2/pull/36/files 
what's going on? 

Can you rebase?

File: view/frontend/templates/forter.phtml [hunk 8]
```diff
@@ -1,9 +1,27 @@
...
-})();
+
+
+
+document.addEventListener('ftr:tokenReady', function(evt) {
+  var xhr = new XMLHttpRequest();
+  xhr.open("POST", '/forter/index/sessions/', true);
+  xhr.setRequestHeader("Content-type", "application/json");
+  xhr.setRequestHeader("X-Requested-With", "XMLHttpRequest");
+  xhr.setRequestHeader("Forter-Token", evt.detail);
```
Comment (orpolaczek) [inline view/frontend/templates/forter.phtml:16]: same. we added this before.

File: view/frontend/templates/forter.phtml (hunk 3 above)
Comment (orpolaczek) [reply view/frontend/templates/forter.phtml:38]: +1

File: view/frontend/templates/forter.phtml [hunk 9]
```diff
@@ -15,8 +33,37 @@ require(["jquery"], function ($) {
     } else {
       console.log('Forter: Invalid Web Id');
     }
+    });
+
+    <?php if ($preAuthIsActive && strlen($elementToObserve) > 1): ?>
+    $(document).on('change', '<?php echo $elementToObserve; ?>', function() {
+      value = $(this).val();
```
Comment (orpolaczek) [inline view/frontend/templates/forter.phtml:40]:
```suggestion
      value = $(this).val() || '';
```

so you won't crash on .substr

**Review by orpolaczek — CHANGES_REQUESTED** [review]
please see my comments

**Review by orpolaczek — APPROVED** [review]

File: etc/adminhtml/system.xml [hunk 10]
```diff
@@ -193,7 +193,39 @@
...
+                    </field>
+                    <field id="enabled_cc_listener_bin" translate="label" type="select" sortOrder="1" showInDefault="1" showInWebsite="1" showInStore="1">
+                        <label>Enable Listener for bin</label>
+                        <source_model>Magento\Config\Model\Config\Source\Yesno</source_model>
+                        <comment>
+                            Enable Javascript Listener For CreditCard Bin
+                        </comment>
+                    </field>
+                    <field id="class_id_identifier" translate="label" type="text" sortOrder="1" showInDefault="1" showInWebsite="1" showInStore="1">
+                      <label>Class or Id identifer for the Listener</label>
```
Comment (erezkoren) [inline etc/adminhtml/system.xml:221]: identifer typo --> `identifier` @pniel-cohen @zachGirit

File: etc/adminhtml/system.xml [hunk 11]
```diff
@@ -193,7 +193,39 @@
...
+                      <comment>
+                          Please add the class or ID of the input of the CC you want to observe.
```
Comment (erezkoren) [inline etc/adminhtml/system.xml:223]: can you add real example

File: view/frontend/templates/forter.phtml [hunk 12]
```diff
@@ -1,9 +1,27 @@
...
+)();
 
 require(["jquery"], function ($) {
   $(function () {
```
Comment (erezkoren) [inline view/frontend/templates/forter.phtml:27]: @pniel-cohen @zachGirit we need to remove the dependency on jQuery, some merchants not supporting it

**Pull request merged**